modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentCommunication.java

```diff
@@ -337,50 +337,63 @@ void sendUndeployRequest(String rsrcName, Collection<ClusterNode> rmtNodes) thro
                 GridIoPolicy.P2P_POOL);
         }
     }
-
+    
+    /**
+     * Queries whether a node is a direct or indirectly responsible for 
+     * the request being processed by this thread.
+     * 
+     * @param  node The node in question
+     * @return true if the node is active
+     */
+    boolean nodeOriginatedCurrentRequest(final ClusterNode node) {
+       // activeReqNodeIds is thread local.
+       Collection<UUID> nodeIds = activeReqNodeIds.get();
+       
+       return (nodeIds != null && nodeIds.contains(node.id()));
+    }
+    
     /**
      * Sends request to the remote node and wait for response. If there is
-     * no response until threshold time, method returns null.
+     * no response until threshold time, method returns null.  The receiver
+     * may forward the request to other nodes, but not to nodes that have been
+     * searched already, or nodes we intend to search directly.
      *
+     * The exclusion list severely weakens the potential n-squared behavior when a 
+     * resource does not exist. If the resource does not exist, all nodes 
+     * in the graph must be touched at least once.   To avoid all multiple touches 
+     * for the resource-not-found case, the response would need to include the 
+     * nodes forwarded to, so they could be added to the exclusion list.  However
+     * that seems like overkill for this kinds of graphs peer class loading will generate.
      *
      * @param rsrcName Resource name.
      * @param clsLdrId Class loader ID.
-     * @param dstNode Remote node request should be sent to.
+     * @param dstNode Remote node request should be sent to. The caller must
+     *      avoid using an originating node.
      * @param threshold Time in milliseconds when request is decided to
      *      be obsolete.
+     * @param exclusionList - Nodes to avoid recursively searching, in addition to 
+     *      the node(s) that forwarded this request initially.
      * @return Either response value or {@code null} if timeout occurred.
      * @throws IgniteCheckedException Thrown if there is no connection with remote node.
      */
     @SuppressWarnings({"SynchronizationOnLocalVariableOrMethodParameter"})
     GridDeploymentResponse sendResourceRequest(final String rsrcName, IgniteUuid clsLdrId,
-        final ClusterNode dstNode, long threshold) throws IgniteCheckedException {
+        final ClusterNode dstNode, long threshold, final Collection<UUID> nodesToSkip) throws IgniteCheckedException {
```

Comment:
Why final? Is this an explicit convention in the Ignite code?

Reply:
Seems pretty inconsistent.